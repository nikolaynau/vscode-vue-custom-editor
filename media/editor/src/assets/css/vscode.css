@import "vscode-theme";
@import "vscode-default";
