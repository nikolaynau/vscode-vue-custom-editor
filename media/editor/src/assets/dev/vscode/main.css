@import 'vscode-default/all';

body {
  background-color: var(--vscode-editor-background);
}
