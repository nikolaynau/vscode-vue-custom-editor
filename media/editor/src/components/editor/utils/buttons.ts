export interface ButtonDefinition {
  id: number;
  label: string;
  value: number;
  side: 'left' | 'right';
}

export function createButtons(): Array<ButtonDefinition> {
  return [
    {
      id: 1,
      label: '-10',
      value: -10,
      side: 'left'
    },
    {
      id: 2,
      label: '-5',
      value: -5,
      side: 'left'
    },
    {
      id: 3,
      label: '-1',
      value: -1,
      side: 'left'
    },
    {
      id: 4,
      label: '+1',
      value: 1,
      side: 'right'
    },
    {
      id: 5,
      label: '+5',
      value: 5,
      side: 'right'
    },
    {
      id: 6,
      label: '+10',
      value: 10,
      side: 'right'
    }
  ];
}
