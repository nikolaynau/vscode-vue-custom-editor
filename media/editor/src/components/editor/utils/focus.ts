export function nextFocusElement(): void {
  const activeElement = document.activeElement;
  if (
    !activeElement?.classList.contains('v-editor__input') &&
    !activeElement?.classList.contains('v-action') &&
    !activeElement?.classList.contains('v-button')
  ) {
    return;
  }
  let nextElement: Element | null | undefined =
    activeElement?.nextElementSibling;
  if (nextElement?.classList.contains('v-editor__controls')) {
    nextElement = nextElement.children[0];
  } else if (
    !nextElement &&
    activeElement?.parentElement?.nextElementSibling?.classList.contains(
      'v-editor__input'
    )
  ) {
    nextElement = activeElement?.parentElement?.nextElementSibling;
  }
  if (!nextElement) {
    nextElement =
      activeElement?.parentNode?.parentNode?.children[0]?.children[0];
  }

  if (nextElement) {
    (nextElement as HTMLElement).focus();
  }
}

export function prevFocusElement(): void {
  const activeElement = document.activeElement;
  if (
    !activeElement?.classList.contains('v-editor__input') &&
    !activeElement?.classList.contains('v-action') &&
    !activeElement?.classList.contains('v-button')
  ) {
    return;
  }
  let prevElement: Element | null | undefined =
    activeElement?.previousElementSibling;
  if (prevElement?.classList.contains('v-editor__controls')) {
    prevElement = prevElement.lastElementChild;
  } else if (
    !prevElement &&
    activeElement?.parentElement?.previousElementSibling?.classList.contains(
      'v-editor__input'
    )
  ) {
    prevElement = activeElement?.parentElement?.previousElementSibling;
  }
  if (!prevElement) {
    prevElement =
      activeElement?.parentNode?.parentNode?.lastElementChild?.lastElementChild;
  }

  if (prevElement) {
    (prevElement as HTMLElement).focus();
  }
}
