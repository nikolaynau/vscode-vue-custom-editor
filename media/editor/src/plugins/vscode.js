export default {
  install(app) {
    const vscode = window?.acquireVsCodeApi?.();
    app.config.globalProperties.$vscode = vscode;
    app.provide("$vscode", vscode);
  }
}
