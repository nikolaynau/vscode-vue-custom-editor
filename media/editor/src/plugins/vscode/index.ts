import type { Plugin, App } from 'vue';
import { vscodeKey } from './injection-symbols';
import vscodeApi from './vscode-api';

export default {
  install(app: App) {
    const vscode =
      (import.meta.env.MODE === 'development' ||
        import.meta.env.MODE === 'test') &&
      typeof window.acquireVsCodeApi !== 'function'
        ? vscodeApi
        : window.acquireVsCodeApi();

    app.config.globalProperties.$vscode = vscode;
    app.provide(vscodeKey, vscode);
  }
} as Plugin;
