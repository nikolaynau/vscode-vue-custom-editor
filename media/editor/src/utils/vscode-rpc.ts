import logger from '@/utils/logger';
import { RpcProvider } from 'worker-rpc';

export class VsCodeRpc {
  public provider: RpcProvider;

  public constructor(vscode: VsCodeApi) {
    this.provider = new RpcProvider(message => vscode.postMessage(message));

    this.handleMessage = this.handleMessage.bind(this);
    this.handleError = this.handleError.bind(this);

    this.provider.error.addHandler(this.handleError);
    window.addEventListener('message', this.handleMessage);
  }

  private handleMessage(e: MessageEvent) {
    this.provider.dispatch(e.data);
  }

  private handleError(err: Error) {
    logger.error('[Rpc]: rpc provider error', err);
  }

  public destroy() {
    window.removeEventListener('message', this.handleMessage);
    const provider = this.provider as any;
    provider._dispatch = null;
    provider._rpcHandlers = [];
    provider._signalHandlers = {};
  }
}
