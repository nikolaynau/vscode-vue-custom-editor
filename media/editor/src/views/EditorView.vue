<script setup lang="ts">
import { toRefs, ref, onBeforeMount, watch } from 'vue';
import VEditor from '@/components/editor/VEditor.vue';
import type { ChangeEvent } from '@/components/editor/composables/use-document-model';
import type { InspectorDataModel } from '@/components/inspector/composables/use-inspector';
import { useVsCode } from '@/plugins/vscode/use-vscode';
import { VsCodeRpc } from '@/utils/vscode-rpc';
import type {
  EditCommandArray,
  RawDocument,
  RawJsonDocument
} from '@/components/editor/utils/types';

export interface Props {
  inspectorEnabled?: boolean;
}

const props = withDefaults(defineProps<Props>(), {
  inspectorEnabled: false
});

const { inspectorEnabled } = toRefs(props);
const editor = ref<InstanceType<typeof VEditor> | null>(null);
const vscode = useVsCode();
const rpc = new VsCodeRpc(vscode);
let pendingInitialData:
  | {
      data: RawDocument;
      editOperations: EditCommandArray;
    }
  | undefined = undefined;

function onChangeValue(event: ChangeEvent) {
  rpc.provider.signal('edit', event);
}

function onUpdateInspector(model: InspectorDataModel) {
  if (inspectorEnabled.value) {
    rpc.provider.signal('updateInspector', model);
  }
}

function handleSetInitialData({
  data,
  editOperations
}: {
  data: RawDocument;
  editOperations: EditCommandArray;
}) {
  if (editor.value) {
    const { setData, applyEdits, sendInspectorModel } = editor.value;
    setData(data);
    applyEdits(editOperations, false);
    sendInspectorModel();
  } else {
    pendingInitialData = { data, editOperations };
  }
}

function handleGetFileData(): RawJsonDocument {
  return editor.value?.toString(true) ?? '';
}

function handleSetFileData(data: RawDocument) {
  editor.value?.setData(data);
}

function handleApplyEdits({
  editOperations,
  notify
}: {
  editOperations: EditCommandArray;
  notify: boolean;
}) {
  editor.value?.applyEdits(editOperations, notify);

  if (!notify) {
    editor.value?.sendInspectorModel();
  }
}

function handleNeedUpdateInspector() {
  editor.value?.sendInspectorModel();
}

rpc.provider.registerRpcHandler('getFileData', handleGetFileData);
rpc.provider.registerRpcHandler('setFileData', handleSetFileData);
rpc.provider.registerRpcHandler('applyEdits', handleApplyEdits);
rpc.provider.registerRpcHandler('setInitialData', handleSetInitialData);
rpc.provider.registerSignalHandler(
  'needUpdateInspector',
  handleNeedUpdateInspector
);

onBeforeMount(() => {
  rpc.destroy();
});

watch(editor, () => {
  if (pendingInitialData && editor.value) {
    handleSetInitialData(pendingInitialData);
    pendingInitialData = undefined;
  }
});
</script>

<template>
  <div class="editor-container">
    <VEditor
      ref="editor"
      keyboard-enabled
      focus-on-start
      :inspector-enabled="inspectorEnabled"
      @change-value="onChangeValue"
      @update-inspector="onUpdateInspector"
    />
  </div>
</template>

<style>
@import 'editor-view';
</style>
