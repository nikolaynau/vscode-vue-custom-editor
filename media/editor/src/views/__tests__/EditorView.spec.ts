import { describe, it, expect, vi, afterEach } from 'vitest';
import { mount } from '@vue/test-utils';
import EditorView from '../EditorView.vue';
import { VsCodeRpc } from '@/utils/vscode-rpc';
import type { PlusValueCommand } from '@/components/editor/utils/types';
import vscode from '@/plugins/vscode';
import type { ChangeEvent } from '@/components/editor/composables/use-document-model';

vi.mock('../../utils/vscode-rpc.ts', () => {
  const VsCodeRpc = vi.fn();
  VsCodeRpc.prototype.provider = {
    signal: vi.fn(),
    registerRpcHandler: vi.fn(),
    registerSignalHandler: vi.fn()
  };
  VsCodeRpc.prototype.destroy = vi.fn();
  return { VsCodeRpc };
});

describe('EditorView', () => {
  afterEach(() => {
    vi.restoreAllMocks();
  });

  it('render default props', () => {
    const wrapper = mount(EditorView, { global: { plugins: [vscode] } });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('send edit signal', async () => {
    const wrapper = mount(EditorView, { global: { plugins: [vscode] } });

    await wrapper
      .find('.v-editor__controls')
      .find('.v-button')
      .trigger('click');

    expect(wrapper.find('.v-editor__input').text()).toBe('-10');

    expect((VsCodeRpc as any).mock.instances).toHaveLength(1);
    expect(
      (VsCodeRpc as any).mock.instances[0].provider.signal
    ).toBeCalledTimes(1);
    expect((VsCodeRpc as any).mock.instances[0].provider.signal).toBeCalledWith(
      'edit',
      {
        changes: [
          {
            applied: {
              name: 'plus',
              payload: { value: -10 }
            } as PlusValueCommand,
            reverse: {
              name: 'plus',
              payload: { value: 10 }
            } as PlusValueCommand
          }
        ],
        versionId: 1
      } as ChangeEvent
    );
  });
});
