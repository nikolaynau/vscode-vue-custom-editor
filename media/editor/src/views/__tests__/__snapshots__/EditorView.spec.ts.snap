// Vitest Snapshot v1

exports[`EditorView > render default props 1`] = `
"<div class=\\"editor-container\\">
  <div class=\\"v-editor\\">
    <div class=\\"v-editor__content\\">
      <div class=\\"v-editor__controls\\"><a class=\\"v-button\\" tabindex=\\"0\\">-10</a><a class=\\"v-button\\" tabindex=\\"0\\">-5</a><a class=\\"v-button\\" tabindex=\\"0\\">-1</a></div>
      <div class=\\"v-editor__input\\" tabindex=\\"0\\">0</div>
      <div class=\\"v-editor__controls\\">
        <div class=\\"v-action\\" title=\\"Clear Counter (Delete)\\" tabindex=\\"0\\"><span class=\\"codicon codicon-clear-all\\"></span></div><a class=\\"v-button\\" tabindex=\\"0\\">+1</a><a class=\\"v-button\\" tabindex=\\"0\\">+5</a><a class=\\"v-button\\" tabindex=\\"0\\">+10</a>
      </div>
    </div>
    <div class=\\"v-editor__shortcuts\\">
      <div class=\\"v-keyboard-shortcuts v-keyboard-shortcuts--label-right\\">
        <dl>
          <dt>Increase 小ounter</dt>
          <dd>
            <div class=\\"v-keybinding-label\\" title=\\"UpArrow\\">
              <!--v-if--><span class=\\"v-keybinding-label__key\\">UpArrow</span>
            </div>
          </dd>
        </dl>
        <dl>
          <dt>Decrease 小ounter</dt>
          <dd>
            <div class=\\"v-keybinding-label\\" title=\\"DownArrow\\">
              <!--v-if--><span class=\\"v-keybinding-label__key\\">DownArrow</span>
            </div>
          </dd>
        </dl>
        <dl>
          <dt>Clear 小ounter</dt>
          <dd>
            <div class=\\"v-keybinding-label\\" title=\\"Delete\\">
              <!--v-if--><span class=\\"v-keybinding-label__key\\">Delete</span>
            </div>
          </dd>
        </dl>
        <dl>
          <dt>Clear 小ounter</dt>
          <dd>
            <div class=\\"v-keybinding-label\\" title=\\"Backspace\\">
              <!--v-if--><span class=\\"v-keybinding-label__key\\">Backspace</span>
            </div>
          </dd>
        </dl>
      </div>
    </div>
  </div>
</div>"
`;
